tests/flytekit/common/workflows/sagemaker.py

```diff
@@ -0,0 +1,192 @@
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import random
+from operator import add
+
+from six.moves import range
+
+from flytekit.sdk.tasks.sagemaker.trainingjob import SdkSimpleTrainingJobTask, DataChannel
+from flytekit.sdk.tasks.sagemaker.hpojob import SdkSimpleHPOJobTask, HPOJobTuningStrategy, HPOJobObjective
+from flytekit.sdk.tasks.sagemaker.parameter_range import IntegerParameterRange, ContinuousParameterRange, CategoricalParameterRange
+from flytekit.sdk.tasks.sagemaker import trainingjob_task
+from flytekit.sdk.tasks import inputs, outputs, python_task
+from flytekit.sdk.types import Types
+from flytekit.sdk.workflow import workflow_class, Input, Output
+import xgboost as xgb
+
+example_hyperparams = {
+    "base_score": "0.5",
+    "booster": "gbtree",
+    "csv_weights": "0",
+    "dsplit": "row",
+    "grow_policy": "depthwise",
+    "lambda_bias": "0.0",
+    "max_bin": "256",
+    "max_leaves": "0",
+    "normalize_type": "tree",
+    "objective": "reg:linear",
+    "one_drop": "0",
+    "prob_buffer_row": "1.0",
+    "process_type": "default",
+    "rate_drop": "0.0",
+    "refresh_leaf": "1",
+    "sample_type": "uniform",
+    "scale_pos_weight": "1.0",
+    "silent": "0",
+    "sketch_eps": "0.03",
+    "skip_drop": "0.0",
+    "tree_method": "auto",
+    "tweedie_variance_power": "1.5",
+    "updater": "grow_colmaker,prune",
+}
+
+simple_xgboost_trainingjob_task = SdkSimpleTrainingJobTask(
+    trainingjob_conf={
+        "InstanceType": "ml.m4.xlarge",
+        "InstanceCount": 1,
+        "VolumeSizeInGB": 25,
+    },
+    stopping_condition={
+        "MaxRuntimeInSeconds": 43200,
+        "MaxWaitTimeInSeconds": 43200
+    },
+    algorithm_specification={
+        "TrainingInputMode": "File",
+        "AlgorithmName": "xgboost",
+        "Version": "0.72",
+    },
+    cache_version='1',
+    cachable=True,
+)
+
+simple_hpojob_task = SdkSimpleHPOJobTask(
+    max_number_of_training_jobs=10,
+    max_parallel_training_jobs=5,
+    cache_version='1',
+    retries=2,
+    cachable=True,
+)
+
+
+def calculate_hyperparameter_max_range(a, b):
+    return a + b
+
+def sample_eval_function(y_predicted, y_true):
+    assert(y_predicted.shape[0] > 0)
+    assert(y_predicted.shape == y_true.shape)
+    err_count = 0
+    for i in enumerate(y_predicted):
+        if y_predicted[i] != y_true[i]:
+            err_count += 1
+
+    return "err rate", err_count/len(y_predicted)
+
+
+@inputs(
+    extra_input1=Types.Integer,
+    extra_input2=Types.Integer,
+)
+@outputs(
+    extra_output1=Types.Blob,
+)
+@trainingjob_task(
+    trainingjob_conf={
+        "InstanceType": "ml.m4.xlarge",
```

Comment:
Customer maybe more inclined to specify cores and memory instead of `InstanceType`, since they may don't know how much resource the instance have.

Reply:
Thanks for the comment! 

I had briefly discussed this with Ketan and Haythem. The problem is that SageMaker currently doesn't support fractional resources (that is, you have to be billed for the entire instance even though your program uses only a part of it). Basically it kind of conflicts with the current semantic in flytekit around resource. So If we allow users to specify  `cpu_limit` and `memory_limit` for SageMaker tasks, users might be misled and found that they are "overcharged" for the resource they didn't claim.

Exposing `InstanceType` and `InstanceCount` to the users is currently the most viable workaround I can think of to address the above issue. 

Any suggestion? :D

Reply:
got it. I don't know the solution either. :D
I was thinking calculate similarity based on user's `cpu_limit` and `memory_limit` between each instance type's. and choose the smallest based on some distance ( maybe x*y/||x||*||y||. (x= (core, memory)) ). but thought it is too complicated and error prone. 

so really i don't know. :(

Reply:
my preference is to keep the interface like Sagemaker has with the instance ID, we can document that (just point to aws documentation) and if they eventually support cpu_limit then we can support it

Reply:
I think one example here is that if a user specifies `cpu_request=5`, but there are only a 4-core option and a 8-core option available among all the instance types. 

In this case, we cannot choose the 4-core option because users specifies a minimum requirement of 5 cores; therefore, we have to choose the 8-core option because that's the smallest instance that satisfies user's requirement. That choice, unfortunately, causes overcharging; the bigger problem is that this overcharging is not easily visible to the users and may causes "surprises" :(